### PR TITLE
Range type

### PR DESCRIPTION
**Pull request by matthuhiggins**
Adds support for elasticsearch's [range datatype](https://www.elastic.co/guide/en/elasticsearch/reference/current/range.html).

### CODE REVIEW

File: lib/elastic_record/as_document.rb [hunk 1]
```diff
@@ -16,27 +16,40 @@ def as_partial_update_document
...
     end
+
+    def value_for_elastic_search_object(object)
+      object.respond_to?(:as_search_document) ? object.as_search_document : object
+    end
+
+    def value_for_elastic_search_range(range)
+      gte = range.begin unless range.begin == -Float::INFINITY
+      lte = range.end unless range.end == Float::INFINITY
+      {'gte' => gte, 'lte' => lte}
```
Comment (dcromer) [inline lib/elastic_record/as_document.rb]: If the beginning or end are infinity, then our document value is `nil`.  Is this acceptable for an ES range?  The documentation does not mention anything about nil values.

Comment (dcromer) [reply lib/elastic_record/as_document.rb]: Ruby ranges can be specified in reverse:  

```
record = SpecialFieldsModel.new(meta: ['whatever'], book_length: 250..0)
```

This results in `{"gte" => 250, "lte" => 0}`.  Is that expected?

Comment (dcromer) [reply lib/elastic_record/as_document.rb]: OK, some more clarity on this:

1) If gte/lte is nil, _or one or both keys are omitted entirely_, the document indexes successfully, and the range is unbounded on that key(s).

2) If gte > lte, the document fails to index.

```
PUT /range_index/_doc/3 
{
  "expected_attendees" : { 
		"gte": 20,
		"lte": 10
  }
}
=> 
{
	"error": {
		"root_cause": [
			{
				"type": "mapper_parsing_exception",
				"reason": "failed to parse [expected_attendees]"
			}
		],
		"type": "mapper_parsing_exception",
		"reason": "failed to parse [expected_attendees]",
		"caused_by": {
			"type": "illegal_argument_exception",
			"reason": "min value (20) is greater than max value (10)"
		}
	},
	"status": 400
}
```

File: lib/elastic_record/as_document.rb [hunk 2]
```diff
@@ -16,27 +16,52 @@ def as_partial_update_document
...
+
+    private
+
+      def ordered_range(range)
+        if range.begin > range.end
+          range.end..range.begin
+        else
+          range
+        end
+      end
```
Comment (matthuhiggins) [inline lib/elastic_record/as_document.rb]: This is silly. Why support this edge case? Is there a use case for it?

Comment (matthuhiggins) [reply lib/elastic_record/as_document.rb]: And if required, you could use min/max with fewer lines of code in the method above :)

Comment (dcromer) [reply lib/elastic_record/as_document.rb]: All elasticsearch ranges are ruby ranges.  Some ruby ranges are elasticsearch ranges.

Should it be possible to initialize a document with values that won't serialize?

> you could use min/max with fewer lines of code in the method above :)

Surprisingly you can't when `begin > end`!  [Range#min](https://ruby-doc.org/core-2.2.0/Range.html#method-i-min)

Comment (matthuhiggins) [reply lib/elastic_record/as_document.rb]: What test case shows this problem?

Comment (dcromer) [reply lib/elastic_record/as_document.rb]: > What test case shows this problem?

https://github.com/data-axle/elastic_record/pull/71/files#diff-32fef10a3489f3c4d2fd0ccf18043c8cR66

File: test/elastic_record/as_document_test.rb [hunk 3]
```diff
@@ -48,17 +50,33 @@ def author
...
-    def meta
-      { some: "value" }
-    end
   end
 
   def test_as_search_document_with_special_fields
-    doc = SpecialFieldsModel.new.as_search_document
+    record = SpecialFieldsModel.new(meta: { some: "value" }, book_length: 250..500)
+
+    doc = record.as_search_document
```
Comment (matthuhiggins) [inline test/elastic_record/as_document_test.rb]: Duplicate tests below. delete this one.

File: lib/elastic_record/as_document.rb (hunk 2 above)
Comment (matthuhiggins) [reply lib/elastic_record/as_document.rb]: I see. Change superstore to make the range nil if begin > end. Allow ES to crash here if the data is bad.

File: lib/elastic_record/as_document.rb [hunk 4]
```diff
@@ -16,27 +16,43 @@ def as_partial_update_document
...
         value
       end
     end
+
+    def value_for_elastic_search_object(object)
+      object.respond_to?(:as_search_document) ? object.as_search_document : object
+    end
+
+    def value_for_elastic_search_range(range)
+      if range.begin <= range.end
```
Comment (matthuhiggins) [inline lib/elastic_record/as_document.rb]: Delete this. Allow it to crash.

Comment (matthuhiggins) [reply lib/elastic_record/as_document.rb]: The crash will never happen though because ORM layer will never create the scenario.

Comment (dcromer) [reply lib/elastic_record/as_document.rb]: elastic_record does not depend on any ORM layer besides active_model.  You are correct that superstore will prevent this, though.

Comment (dcromer) [reply lib/elastic_record/as_document.rb]: removed in 8fc5191ec2216707740d5bd5148bc0dd9e158080

tests reflect expected behavior

**Pull request merged**